Support model metadata

pritamdamania87 opened the pull request:
Summary: Adding a generic string `metadata` field as part of the model to capture additional metadata with the model.

Differential Revision: D13579029

zdevito commented on torch/csrc/jit/script/module.h:
```diff
@@ -563,6 +563,14 @@ struct Module {
     }
   }
 
+  void set_metadata(const std::string& metadata) {
```
This API is not symmetric: on save metadata is taken from an argument, but on load it is made to be part of the module. This is weird: meta-data is stored on the model but it is getting stored on a Module on load. Then a bunch of submodule will have meta-data that is unused. This is a situation that we wanted to avoid by putting meta-data at the module level rather than on each module. Can the load function also just gain a std::string* that will be where it puts metadata if found?

zdevito commented on torch/csrc/jit/import.cpp:
```diff
@@ -196,8 +210,13 @@ at::Tensor ScriptModuleDeserializer::loadTensor(
+    const string& model_metadata,
+    bool main_module) {
   std::shared_ptr<script::Module> module = moduleLookup_(moduleStack_);
+  if (main_module) {
```
Boolean flag is weird here. Probably because it is mixing up the purpose of a Module and a Model.

zdevito reviewed (commented):
Code changes are mostly good, but I do not think Module objects should gain a meta-data property when the meta-data is being stored on disk in the Model.

apaszke reviewed (changes requested):
Honestly this is somewhat weird. Can't we simply make it possible for people to add attributes to `ScriptModule`s, even if they will never end up being used in their compiled code? That seems like a more natural extension and avoids duplicating the API in an awkward way.

pritamdamania87 commented:
@apaszke This metadata needs to be saved in the model file and we should be able to access this metadata when we load the model in Python and C++. Is your suggestion to build a generic framework where users can set any number attributes on ScriptModule, we serialize these attributes to the model file, deserialize these attributes on model load and finally allow users to access these attributes from Python and C++?

apaszke commented:
Yeah. I agree it might be slightly awkward in C++, but it results in a very natural API in Python. You just assign attributes to it, and they are preserved when you load it.

zdevito commented:
I think we should merge this as is. I don't want to build metadata deeply into the Modules until we have a better idea of what particular metadata does. Metadata tends to be a grab bag of random functionality. 'Good' metadata ends up just being promoted to actual API in Module. Forcing all attributes to be loaded separately from the modules ensures that it is easy to understand how the attributes are being used  because to preserve the meta-data the caller of save must provide it (meaning we can track where it was generated from), and then must pipe it to somewhere after load (meaning we can track where they go to).

apaszke commented:
I don't get it. How is this weird packaging on the side, with completely separate save/load paths, better than having support for first-class attributes? It's just a super complicated API with little explanation. If one just wants to keep both metadata and models why can't you just have two files side by side (and correlate them by name), or simply append to whatever you got from `.save()`?

apaszke commented:
Like, we already have `torch.load`, `torch.jit.load`, and `torch::load`. All of those are incompatible between them, and this patch makes the APIs diverge even further. We've already made a mess, so let's not make it worse.

zdevito commented:
I talked with @apaszke offline. We think this can be tweaked a bit to make it more sustainable. Rather than have a single meta-data file, we can allow users of the API to store additional files in the zip archive that can be used to store meta-data.  Accessing this data can be done simply by treating the archive as a zip file and accessing it using standard zip tools. However, we will also provide helper functions to do this packaging using the save/load APIs. 

In python this will be a kwarg called `_extra_files` to save. It is a dictionary from filename to file contents (both strings):
```
my_module.save("filename", _extra_files= { 'predictor_metadata.json': 'my_metadata'})
```
Then the loader says what meta-data files they want to load as well:
```
# loader declares what extra files it wants to load, to ensure we always know what meta-data is expected to be there.
meta_data = { 'predictor_metadata.json': None }
model = torch.jit.load("filename", _extra_files = meta_data)
assert(meta_data['predictor_metadata.json' == 'my_metadata')
```
The API in C++ can be analogous. Extra files will go in their own folder in the zip archive 'extra/' to make sure they never overlap with files we add to the format.

This allow different producers/consumers of metadata to be able to work side-by-side without having to agree on formats.

dzhulgakov commented:
Nice resolution (funny thing - I was suggesting similar generalization to @pritamdamania87 but we decided to to overcomplicate it :) )

pritamdamania87 commented:
@zdevito @apaszke I've updated the PR as per the suggestion mentioned above.

zdevito commented on torch/csrc/jit/export.cpp:
```diff
@@ -540,14 +547,22 @@ void ScriptModuleSerializer::serialize(const script::Module& module) {
+  // Write out extra files.
+  for (const auto& kv : extra_files) {
+    const std::string key = "extra/" + kv.first;
+    model_def->add_extra_files()->set_key(key);
```
I don't think we need to add an entry to torch.proto here. Avoiding it would allow other tools to just stash files in the extra folder without also having to edit the data in the model_def. We expect some users of meta-data will end up just adding files to the zip archive this way.

zdevito commented on torch/csrc/jit/import.cpp:
```diff
@@ -112,6 +114,26 @@ void ScriptModuleDeserializer::deserialize(
   const auto& module_def = model_def.main_module();
+
+  // Load extra files.
+  for (const auto& kv : extra_files) {
```
This would be as simple as this if we removed the entry in model_def:
```
reader_.getRecord("extra/" + kv.first)
```

dzhulgakov replied:
I guess we store the size to make it mmapable. Though for extra files it might not be necessary (and probably can be extracted from zip format anyway)

zdevito reviewed (approved):
Thanks! I don't think we need the RecordRef for the extra files (see explanation below), but this otherwise looks good to go.

dzhulgakov commented on torch/csrc/jit/import.h:
```diff
@@ -17,28 +17,34 @@ namespace jit {
 using ModuleLookup = std::function<std::shared_ptr<script::Module>(
     const std::vector<std::string>&)>;
 
+static script::ExtraFilesMap default_extra_files;
```
might be easier to just pass pointer instead of reference?

pritamdamania87 replied:
Pointers are messy in pybind (needs to be smart pointer).

dzhulgakov commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -619,23 +624,31 @@ FunctionSchema getSchemaWithNameAndDefaults(
 void initJitScriptBindings(PyObject* module) {
   auto m = py::handle(module).cast<py::module>();
 
+  py::bind_map<ExtraFilesMap>(m, "ExtraFilesMap");
```
so we need to do this because containers are not mutable by default? maybe add a comment?

pritamdamania committed:
PyTorch model metadata. (#16275)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/16275

Adding a generic string `metadata` field as part of the model to capture additional metadata with the model.

Reviewed By: dzhulgakov

Differential Revision: D13579029

fbshipit-source-id: 82b9a018d798594953254bb0911ef89e2885c86f
